### PR TITLE
Add JAX VMEC, Boozer, and virtual-casing wrappers

### PR DESCRIPTION
**Pull request by rogeriojorge**
## Summary

This draft PR adds Simsopt integration for the JAX MHD stack:

- `VmecJax` wrapper and VMEC-JAX diagnostics, including fixed-boundary runs, profile support, quasisymmetry residuals, exact optimization helpers, and boundary-field tangent columns.
- `BoozerJax` wrapper around `booz_xform_jax`, with Simsopt-style accessors/tests and direct use of the merged upstream vectorized API.
- `VirtualCasingJax` wrapper around `virtual_casing_jax`, including normal-field evaluation, JVP/Jacobian helpers, save/load/plot parity, and direct use of the merged upstream normal-field JVP column API.
- JAX versions of the requested intermediate and advanced examples, including the finite-beta single-stage example with a `CI=true` smoke path and environment-variable controls for bounded local regression runs.
- Documentation page and validation plots for VMEC-JAX, Booz-Xform-JAX, virtual-casing-JAX, finite-beta derivative timings, and reduced finite-beta legacy/JAX parity.
- Public CI now installs the public upstream `vmec_jax` and `booz_xform_jax` packages. It attempts `virtual_casing_jax` opportunistically, but that upstream repository is currently private, so virtual-casing JAX tests skip in public CI unless credentials are available.
- Conda CI now calls the `conda-build` executable directly instead of relying on `conda build` plugin discovery, which currently fails on the macOS runner with conda 26.
- An opt-in legacy VMEC comparison test is available for local environments with legacy VMEC2000 and virtual-casing installed.

The repository uses `master` as the default branch and does not currently expose a `main` branch, so this PR targets `master`.

## Upstream Dependencies

The Simsopt wrappers now depend on upstream APIs that have already been merged:

- `uwplasma/vmec_jax` PR #8, merged as `aa034d3`.
- `uwplasma/booz_xform_jax` PR #1, merged as `a3383c6`.
- `uwplasma/virtual_casing_jax` PR #1, merged as `d3a9060`.

The Simsopt wrapper code calls those merged APIs directly instead of carrying fallback implementations.

## Validation

Local validation run on this branch:

- `.venv/bin/python -m pytest tests/mhd/test_jax_examples.py tests/mhd/test_vmec_jax.py tests/mhd/test_boozer_jax.py tests/mhd/test_vmec_diagnostics_jax.py tests/mhd/test_virtual_casing_jax.py tests/mhd/test_vmec_jax_legacy_comparison.py -q`
  - latest focused-suite run: `27 passed`, `1 skipped`, `6` example subtests passed, one upstream JAX warning, wall time `44.80s`
- `SIMSOPT_RUN_LEGACY_VMEC_JAX_COMPARISON=1 JAX_ENABLE_X64=1 .venv_compare_py312/bin/python -m pytest tests/mhd/test_vmec_jax_legacy_comparison.py -q`
  - `1 passed in 7.24s`
- `.venv/bin/python -m pytest tests/mhd/test_jax_examples.py::JaxExamplesTests::test_single_stage_finite_beta_jax_ci_smoke -q`
  - `1 passed in 30.30s`
- `CI=true MPLBACKEND=Agg .venv/bin/python examples/3_Advanced/single_stage_optimization_finite_beta_jax.py`
  - completed in about 22 seconds with `J=269.4487` and `|grad|=2.1079e+03`
- `.venv/bin/python -m sphinx -b html docs/source docs/build/html`
  - build succeeded with the pre-existing C++ API documentation warnings

The finite-beta CI smoke test parses and bounds the exact-gradient smoke output:

- objective: `200 < J < 350`
- gradient norm: `1e3 < |grad| < 5e3`

A non-CI `12 x 12`, `2 + 2` finite-beta single-stage run was also attempted using the new `SIMSOPT_JAX_QUICK=true` controls. It progressed through setup and stage-II artifacts, but was still inside the exact single-stage gradient path after more than six minutes, so I stopped it. That longer trajectory comparison remains a deliberate benchmark item rather than a routine PR smoke test.

## Legacy Comparison Status

I created a separate local Python 3.12 comparison environment with legacy VMEC2000 and virtual-casing installed. Homebrew Python 3.12 needed an Expat dynamic-library fix, and legacy VMEC needed a venv-local f90wrap handle shim because its generated accessors expect a two-integer empty handle while f90wrap 0.3 reports four integers on this macOS arm64 stack.

Reduced `input.QH_finitebeta`, `12 x 12` comparison results from the refreshed comparison run:

- VMEC aspect relative difference: `3.8e-15`
- VMEC mean-iota relative difference: `2.0e-14`
- VMEC external-current relative difference: `1.4e-13`
- Three-surface QS total relative difference for the finite-beta example target helicity `helicity_n=-1`: `6.6e-6`
- Same three-surface QS check for `helicity_n=+1`: `2.1e-6`
- Legacy virtual-casing vs `VirtualCasingJax` on the same legacy VMEC field: relative L2 `3.0e-4`
- Full `Vmec -> VirtualCasing` vs `VmecJax -> VirtualCasingJax` normal field: relative L2 `2.5e-3`
- Full-inner finite-beta objective relative difference at initial surface/coils: about `1.5e-7`
- Exact JAX finite-beta surface gradient vs a central finite difference of the JAX target/objective path: `5.7e-4` relative directional mismatch

The legacy finite-difference surface gradient and exact JAX surface gradient are not identical: at this reduced resolution the relative L2 difference is about `2.6e-1`. Decomposition shows the fixed-target local `SquaredFlux` geometry derivative agrees with finite differences to about `1.3e-3`; the difference is in the derivative of the VMEC/virtual-casing target field. This is documented as an expected optimizer-trajectory difference rather than hidden.

## Remaining Work Before Ready For Review

- Monitor and fix CI after the private-dependency workflow correction and Conda command fix.
- Decide whether to keep the JAX package installs in the main workflow or move them into a dedicated JAX smoke workflow with optional credentials for `virtual_casing_jax`.
- Run longer, non-smoke finite-beta single-stage optimization regressions and compare trajectories/objective histories.
- Decide which JAX examples should be regular tests, smoke tests, or nightly tests.
- Review feature parity gaps that are intentionally not finished yet, especially free-boundary/asymmetric edge cases and any legacy VMEC/Boozer/virtual-casing diagnostics not covered by the new wrappers.
- Review installation docs and version pins for `vmec_jax`, `booz_xform_jax`, and `virtual_casing_jax`.

### CODE REVIEW

**Comment by rogeriojorge** [comment]
Status update: please keep this PR in draft and do not review yet. We are making virtual_casing_jax public, adding release/coverage CI there, syncing the latest vmec_jax and booz_xform_jax main changes locally, and preparing a research-grade coverage plan plus a production-like finite-beta benchmark before requesting review. Codecov is required to be green before review.

**Comment by rogeriojorge** [comment]
Status update: virtual_casing_jax is now public and its main CI is green. This PR is still draft and should not be reviewed yet. Latest pushed Simsopt head is 7439f60d with public vmec_jax/booz_xform_jax/virtual_casing_jax CI installs, finite-beta benchmark artifact generation, and docs plots. Local validations passed for the focused JAX smoke, docs build, and bounded finite-beta benchmarks: max_mode=1 completed in 26.69 s with squared flux 0.4282 -> 0.05224 and exact single-stage check J=61.2873, |grad|=1.7263e3; max_mode=2 completed in 77.99 s with |grad|=2.1707e3. Codecov is still the required gate before review; the latest CI/Codecov for 7439f60d is pending.

**Comment by rogeriojorge** [comment]
Status update: keep this PR in draft and do not review yet. Since the previous update, `virtual_casing_jax` v0.0.2 was released through trusted PyPI publishing after current-head normal CI/docs/coverage and CI-Large both passed. The finite-beta JAX BFGS native abort has also been fixed in this PR by guarding unsafe line-search trial steps before entering VMEC-JAX/virtual-casing-JAX; the non-CI guarded run completed in 86.01 s with accepted objective evaluations `61.2786` and `59.1642` after the initial `J=61.2873` evaluation. Local focused JAX tests and docs builds are green. The remaining review gate is Codecov: patch coverage is still expected to need targeted tests, especially for `src/simsopt/mhd/virtual_casing_jax.py`, before this PR should move out of draft.

**Comment by rogeriojorge** [comment]
Status update: keep this PR in draft. I added focused, physically meaningful wrapper coverage tests for the JAX MHD layer and pushed them through `a081e5a8`. Local focused validation now passes for `test_boozer_jax.py`, `test_vmec_jax.py`, `test_vmec_diagnostics_jax.py`, `test_virtual_casing_jax.py`, and `test_jax_examples.py`: 44 tests plus 6 subtests in 58.42 s. A combined local coverage run over those tests gives 96% total coverage across the JAX wrapper files: `boozer_jax.py` 98%, `virtual_casing_jax.py` 97%, `vmec_jax.py` 96%, and `vmec_diagnostics_jax.py` 92%. The latest GitHub checks for `a081e5a8` are still queued/in progress, so the remaining gate is the hosted CI/Codecov result.

**Comment by rogeriojorge** [comment]
Status update: pushed `8f1dd77f` to use the released PyPI package `virtual-casing-jax>=0.0.2` in Simsopt public CI and updated the JAX MHD docs accordingly. Local validation after installing the PyPI package shows `virtual-casing-jax` version `0.0.2` from site-packages and `tests/mhd/test_virtual_casing_jax.py` passes (`14 passed`). Local docs build also succeeds with the pre-existing Doxygen/Sphinx warnings. Hosted CI for `8f1dd77f` is still queued/in progress, so this remains draft pending GitHub CI and Codecov.

**Comment by rogeriojorge** [comment]
Status update: final local production-like finite-beta JAX benchmark completed and is now documented in this PR at `b1635786`. Run settings: `virtual-casing-jax==0.0.2`, `SIMSOPT_JAX_MAX_MODE=2`, `12 x 12` VMEC/virtual-casing grids, two stage-II iterations, one guarded BFGS single-stage iteration, and `SIMSOPT_JAX_SINGLE_STAGE_CHECK_ONLY=false`. Runtime was 174.80 s. Stage-II reduced `Jf` from 0.428160 to 0.055790; guarded BFGS evaluated `J=61.2873`, rejected one unsafe line-search trial through the step guard, then accepted `J=61.2626` and `J=55.7930`; final squared flux was 0.0466793. Updated docs plots: `docs/source/jax_mhd_finite_beta_benchmark_history.png` and `docs/source/jax_mhd_finite_beta_benchmark_bnormal.png`. Local docs build passes with the pre-existing warnings. This remains draft until hosted CI and Codecov finish on latest head `b1635786`.

**Comment by rogeriojorge** [comment]
Status update: pushed `7f45d200` with the objective-composition and review-hardening pass. The JAX optimization examples now build stage-I objectives through user-editable `make_stage1_residual_terms` functions and the shared `simsopt.mhd.make_vmec_jax_residuals_from_terms` helper, so users can swap in any JAX-compatible VMEC-state residual without changing the optimizer plumbing. The default terms still use the optimized vmec_jax residual factory, preserving the exact adjoint path.\n\nAlso addressed the Copilot comments: replaced the new runtime `assert` with `ValueError`, bounded the virtual-casing JAX JIT cache and added `clear_virtual_casing_jax_cache`, pinned CI installs to `vmec_jax@e236ba559961daf3b788f00b2443d0fa8cb6ec87` and `booz_xform_jax@809443cad9efc02cd354188aefd1735bc1b4bb4c`, and made the JAX smoke-test timeout/timing gate configurable.\n\nLocal validation on the updated upstream vmec_jax main SHA: `git diff --check` passed; changed files compile; focused JAX MHD tests passed with `47 passed, 6 subtests passed in 61.71s`; docs build passed with the existing 28 warnings. The PR remains draft while hosted CI and Codecov rerun on this new head.

**Comment by rogeriojorge** [comment]
Update pushed: `a000d3d8 Make JAX VMEC objectives modular`.\n\nThis addresses the objective-composition concern by moving reusable pieces into source code and making the examples follow the standard SIMSOPT tuple workflow instead of hard-coded/local residual factories:\n\n- Added `AspectRatioJax` and `VmecJaxLeastSquaresProblem.from_tuples` so exact VMEC-JAX objectives are assembled as separate `(function, target, weight)` terms.\n- Added `QuasisymmetryRatioResidualJax.residuals_from_state` for VMEC-side QS exact optimization.\n- Added `BoozerQuasisymmetryResidualJax` so the Boozer exact residual is reusable from `simsopt.mhd` instead of defined inside the example.\n- Updated the QH and single-stage JAX examples to show aspect ratio and quasisymmetry as separate objective terms, then combine them with `VmecJaxLeastSquaresProblem.from_tuples`.\n- Updated the JAX-MHD docs and added focused tests for the new modular source APIs.\n\nLocal validation:\n- `py_compile` for changed modules/examples/tests: passed.\n- `git diff --check`: passed.\n- Focused modular objective tests: passed.\n- Full JAX MHD slice: `51 passed, 1 warning, 6 subtests passed in 59.01s`.\n- Sphinx docs build: passed with the same existing 28 C++ API/Doxygen warnings.\n\nPR remains draft pending hosted CI/Codecov.

**Comment by rogeriojorge** [comment]
Final upstream-wrapper performance pass pushed: `a5116310 Use latest VMEC JAX performance path`.\n\nWhat changed:\n- Fast-forwarded local `vmec_jax_simsopt` to current upstream main `4e16b40` and updated SIMSOPT CI to install that SHA. This picks up residual-callback JIT caching, QS angle-grid caching, and the optimizer change that reuses the already-computed residual vector for QS history/objective reporting.\n- Confirmed `vmec_jax` PR #8, `booz_xform_jax` PR #1, and `virtual_casing_jax` PR #1 are all merged into their upstream main branches. No new upstream branches/PRs are needed for SIMSOPT functionality.\n- Rechecked ownership boundaries: upstream repos own numerical kernels/JAX functional APIs; SIMSOPT only owns wrapper glue, SIMSOPT object interfaces, tuple objective composition, array convention conversion, legacy-compatible attributes/file format behavior, and coil/surface derivative composition.\n- Removed a small SIMSOPT-side duplication by factoring VMEC-JAX initial-state/signgs setup into one private helper used by both VMEC-side and Boozer-side state objectives.\n\nLocal validation:\n- SIMSOPT focused helper tests: `3 passed`.\n- SIMSOPT full JAX MHD slice against `vmec_jax@4e16b40`: `51 passed, 1 warning, 6 subtests passed in 70.99s`.\n- Upstream `vmec_jax` smoke slice: `45 passed, 1 skipped`.\n- Upstream `booz_xform_jax` smoke slice: `10 passed`.\n- Upstream `virtual_casing_jax` smoke slice: `6 passed`.\n- `py_compile` and `git diff --check`: passed.\n\nCurrent assessment: SIMSOPT contains the necessary wrapper/orchestration layer only; no duplicate core numerical functionality remains that should be moved upstream before review.

File: src/simsopt/mhd/vmec_diagnostics_jax.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        try:
+            self.surfaces = list(surfaces)
+        except TypeError:
+            self.surfaces = [surfaces]
+
+        if weights is None:
+            self.weights = np.ones(len(self.surfaces))
+        else:
+            self.weights = weights
+        assert len(self.weights) == len(self.surfaces)
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply src/simsopt/mhd/vmec_diagnostics_jax.py]: Addressed by replacing the assertion with an explicit ValueError when weights and surfaces have different lengths.

File: src/simsopt/mhd/virtual_casing_jax.py [hunk 2]
```diff
@@ -0,0 +1,699 @@
...
+__all__ = [
+    "VirtualCasingJax",
+    "B_external_normal_from_data",
+    "B_external_normal_jvp_from_data",
+    "B_external_normal_jacobian_from_surface",
+    "local_squared_flux_surface_gradient",
+]
+
+_B_EXTERNAL_JVP_COLUMNS_CACHE = {}
+
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply src/simsopt/mhd/virtual_casing_jax.py]: Addressed by making the virtual-casing JAX helper cache bounded/LRU and adding clear_virtual_casing_jax_cache().

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -111,6 +111,12 @@ jobs:
     - name: Install virtual_casing
       run: pip install -v git+https://github.com/hiddenSymmetries/virtual-casing
 
+    - name: Install public JAX MHD packages
+      run: |
+        pip install -v git+https://github.com/uwplasma/vmec_jax
+        pip install -v git+https://github.com/uwplasma/booz_xform_jax
+        pip install -v "virtual-casing-jax>=0.0.2"
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply .github/workflows/tests.yml:118]: Addressed by pinning vmec_jax and booz_xform_jax installs to known-good commit SHAs and using released virtual-casing-jax from PyPI.

File: tests/mhd/test_jax_examples.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        start = time.perf_counter()
+        try:
+            result = subprocess.run(
+                [sys.executable, example],
+                cwd=ROOT,
+                env=env,
+                capture_output=True,
+                text=True,
+                timeout=60,
+            )
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply tests/mhd/test_jax_examples.py:75]: Addressed by making the JAX example timeout configurable via SIMSOPT_JAX_EXAMPLE_TIMEOUT with a more generous default.

File: tests/mhd/test_jax_examples.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+            result.stdout,
+        )
+        self.assertIsNotNone(match)
+        objective = float(match.group(1))
+        gradient_norm = float(match.group(2))
+        self.assertGreater(objective, 200.0)
+        self.assertLess(objective, 350.0)
+        self.assertGreater(gradient_norm, 1.0e3)
+        self.assertLess(gradient_norm, 5.0e3)
+        self.assertLess(elapsed, 60.0)
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply tests/mhd/test_jax_examples.py]: Addressed by removing the unconditional wall-time assertion; timing is only enforced when SIMSOPT_JAX_EXAMPLE_TIMING_LIMIT is explicitly set.

**Comment by rogeriojorge** [comment]
Ready for review.\n\nFinal cleanup/status:\n- PR has been moved out of draft.\n- All Copilot review threads have been replied to and resolved.\n- The final upstream-wrapper pass is included in `a5116310`, using current `vmec_jax@4e16b40` for the cached residual callback/performance path.\n- Local validation remains green: SIMSOPT full JAX MHD slice `51 passed, 1 warning, 6 subtests passed`; upstream smoke slices passed for `vmec_jax`, `booz_xform_jax`, and `virtual_casing_jax`.\n- CI/CD check: all completed hosted checks visible at this time are passing, with no failures visible; several long-running hosted unit/integration jobs are still in progress after the last push.\n\nImplementation assessment: SIMSOPT contains only the wrapper/orchestration layer needed for this integration; the numerical kernels and differentiable physics remain in the upstream JAX packages.

**Comment by mbkumar** [comment]
@rogeriojorge  How does this PR relate to #604?

**Comment by rogeriojorge** [comment]
vmec_jax had substantial changes since this PR (and renamed to VMEX), so I'll close this PR for now. This needs a new PR with the new changes.
